openeogeotrellis/integrations/zookeeper.py

```diff
@@ -0,0 +1,155 @@
+        8      16.4135469	  17.41354692	           47.29004252
+        9      27.9030297	  28.903029764	           76.193072284
+        10     47.4351505	  48.4351505988	          124.6282228828
+
+        """
+        kz_retry = KazooRetry(max_tries=tries, delay=delay, backoff=backoff)
+        self._logger = logger
+        self._zkhosts = hosts
+        self._zk = self._KazooClient(
+            hosts=hosts,
+            connection_retry=kz_retry,
+            command_retry=kz_retry,
+            timeout=attempt_timeout,
+            logger=logger,
+            read_only=True,
+        )
+        self._zk.start(timeout=total_timeout)
+
+        # Create KazooClient methods:
+        for kazoo_method in ["get", "get_children", "delete", "create", "set"]:
+            setattr(self, kazoo_method, self._make_retrying_kazooclient_method(kazoo_method))
+
+    # We add the kazoo methods just to avoid IDEs getting confused and giving unresolved attribute warnings
+    # but these are not the actual methods as those will be created upon initialization.
+    get = KazooClient.get
+    get_children = KazooClient.get_children
+    delete = KazooClient.delete
+    create = KazooClient.create
+    set = KazooClient.set
+
+    def __del__(self):
+        """
+        The lifetime of the connection is tied to the object. So if the object goes out of scope the connection will be
+        cleaned up as well.
+        """
+        # noinspection PyBroadException
+        try:
+            self._zk.stop()
+        except Exception:
+            pass  # it will stop eventually
+        finally:
+            # noinspection PyBroadException
+            try:
+                self._zk.close()
+            except Exception:
+                pass  # As long as it is used as documented this is fine, risk of connection leakage is very low
+
+    class NoActiveConnectionException(Exception):
+        """This is a helper exception to signal that there is no Active connection at this time."""
+
+    def _wait_100ms_for_zookeeper_connection(self) -> None:
+        """
+        If no active zookeeper connection wait 100 milliseconds and see if active yet.
+        The checking whether the connection is active is a cheap check, so it can be retried aggressively.
+
+        raise ZookeeperClient.NotActiveYetException if at the end no active connection is available.
+        """
+        if not self._zk.connected:  # If already active do not wait
+            sleep(0.1)
+            if not self._zk.connected:
+                if self._logger is not None:
+                    self._logger.debug(f"No active connection for zookeeper client ({self._zkhosts})")
+                raise ZookeeperClient.NoActiveConnectionException()
+
+    _wait_up_to_1minute_for_zookeeper_connection = retry(
+        exceptions=(NoActiveConnectionException,),
+        tries=600,
+        logger=None,
+    )(_wait_100ms_for_zookeeper_connection)
```

Comment:
while reviewing this I got a bit lost in this `_wait_100ms_for_zookeeper_connection`/`_wait_up_to_1minute_for_zookeeper_connection` construction.

Can't it be simplified to a single helper like 

```python
def _wait_on_connected(self, timeout: float=60, sleep=0.1):
    end = time.time() + timeout
    while time.time() < end:
        if self._zk.connected:
            return
        if self._logger is not None:
            self._logger.debug(f"No active connection for zookeeper client ({self._zkhosts})")
        time.sleep(sleep)
    raise ZookeeperClient.NoActiveConnectionException()
```
    



(PS: I would avoid hardcoding implementation details like `100ms` and `1minute` in the function names to make it easier to fine-tune these when necessary)